### ISSUE TITLE
`bootctl install` fails on existing EFI/Boot directory

### ISSUE THREAD

**Issue by martinpitt** [systemd/systemd#3740]
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

> 230
### Used distribution

> Debian unstable

Debian's installer creates a directory `EFI/BOOT` on the EFI partition. `bootctl install` stumbles over that due to vfat's "BOOT" and "Boot" are the same but at the same time aren't" weirdness:

```
# bootctl --path /boot/efi install
Created "/boot/efi/EFI/systemd".
Created "/boot/efi/loader".
Created "/boot/efi/loader/entries".
Copied "/usr/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/efi/EFI/systemd/systemd-bootx64.efi".
```

strace says:

```
mkdir("/boot/efi/EFI/Boot", 0700)       = -1 EEXIST (File exists)
[...]
open("/boot/efi/EFI/Boot/BOOTX64.EFI~", O_WRONLY|O_CREAT|O_EXCL|O_TRUNC|O_CLOEXEC, 0666) = -1 ENOENT (No such file or directory)
```

I can rename `Boot` to `b` and then `b` to `BOOT`, or remove the directory altogether -- in both cases it then succeeds.

Debian mounts `/boot/efi` with `shortname=mixed`, which is the default/recommended mode:

```
          mixed  Display  the  short name as is; store a long name when the short name is not all upper case.
                 This mode is the default since Linux 2.6.32.
```

I'm not entirely sure how to deal with this. The `mkdir("Boot")` is doomed to fail if `BOOT` already exists; when installing bootx64.efi it could try a couple of different capitalizations, like "boot", "Boot", and "BOOT" -- I don't think others are realistic.

**Comment by keszybz** [comment]
Hm, this is strange. I cannot reproduce this here:

```
[root@laptop ~]# dd if=/dev/zero bs=1M count=1 of=/var/tmp/img
1+0 records in
1+0 records out
1048576 bytes (1.0 MB, 1.0 MiB) copied, 0.00188064 s, 558 MB/s
[root@laptop ~]# mkfs.vfat /var/tmp/img
[root@laptop ~]# mkdir /tmp/mount
[root@laptop ~]# mount /var/tmp/img /tmp/mount -o shortname=mixed
[root@laptop ~]# mkdir /tmp/mount/EFI/BOOT -vp
mkdir: created directory '/tmp/mount/EFI'
mkdir: created directory '/tmp/mount/EFI/BOOT'
[root@laptop ~]# mkdir /tmp/mount/EFI/Boot
mkdir: cannot create directory ‘/tmp/mount/EFI/Boot’: File exists
[root@laptop ~]# touch /tmp/mount/EFI/Boot/a
[root@laptop ~]# touch /tmp/mount/EFI/Boot/BOOTX64.EFI~
```

This is 4.5.6-300.fc24.x86_64. What kernel version do you have?
Tthings failing like in your strace seems like a kernel bug.

**Comment by AlexanderKurtz** [comment]
While not directly related to this bug, I suggest to take the opportunity to check that `bootctl` always uses the canonical spelling for file paths on the ESP as defined by the specification [0]. Trying to create the directories/files with any other spelling seems like a bug of its own to me.

[0]  [UEFI Specification](http://www.uefi.org/specifications), version 2.6, section 3.5.1.1

**Comment by keszybz** [comment]
This would mean changing EFI/Boot to EFI/BOOT. Anything else?

**Comment by AlexanderKurtz** [comment]
Yes, for some reason the filenames for the fallback bootloader (`BOOT*.EFI`) defined by the spec are all uppercase, **except** for x64 which uses `BOOTx64.EFI`. No idea what's the reason for this, but according to the strace from @martinpitt, `bootctl` wrongly uses the all-uppercase-variant there as well (lower-case vs. upper-case X).

**Comment by poettering** [comment]
> mkdir("/boot/efi/EFI/Boot", 0700) = -1 EEXIST (File exists)
> [...] open("/boot/efi/EFI/Boot/BOOTX64.EFI~", O_WRONLY|O_CREAT|O_EXCL|O_TRUNC|O_CLOEXEC, 0666) = -1 ENOENT (No such file or directory)

If you check the sources you'll see that we really don't care at all if the dir already exists. we try to create it. if it succeeds, great. If it fails with EEXIST, also great. 

On your system creating "/boot/efi/EFI/Boot" fails, suggesting the dir already exists. But when we create a file in it, then it claims that the dir doesn't exist after all. How can that be? why would the kernel be fine with the upper/lower case differences on mkdir(), but complain when they are in the initial parts of an open() path? there's something very fishy here. What exactly are your mount options for the EFI partition?

**Comment by poettering** [comment]
I just tried this here:

``` sh
[root@sigma systemd]# mkdir /boot/efi/testtest
[root@sigma systemd]# mkdir /boot/efi/TESTTEST
mkdir: cannot create directory ‘/boot/efi/TESTTEST’: File exists
[root@sigma systemd]# touch /boot/efi/testtest/miep
[root@sigma systemd]# touch /boot/efi/TESTTEST/muep
```

This shows that this works correctly here. The mount options used are: `rw,relatime,fmask=0077,dmask=0077,codepage=437,iocharset=ascii,shortname=winnt,errors=remount-ro`

**Comment by poettering** [comment]
it would be good to learn how precisely this works for you...

**Comment by AlexanderKurtz** [comment]
@poettering:

# Environment

```
root@shepard:~# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux stretch/sid"
NAME="Debian GNU/Linux"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
root@shepard:~# systemctl --version
systemd 230
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN
root@shepard:~# systemctl cat boot.mount
# /run/systemd/generator.late/boot.mount
# Automatically generated by systemd-gpt-auto-generator

[Unit]
Description=EFI System Partition Automount
Documentation=man:systemd-gpt-auto-generator(8)
Requires=systemd-fsck@dev-sda2.service
After=systemd-fsck@dev-sda2.service

[Mount]
What=/dev/sda2
Where=/boot
Type=vfat
Options=umask=0077,noauto,rw
root@shepard:~# mount | grep boot
systemd-1 on /boot type autofs (rw,relatime,fd=37,pgrp=1,timeout=120,minproto=5,maxproto=5,direct,pipe_ino=11925)
/dev/sda2 on /boot type vfat (rw,relatime,fmask=0077,dmask=0077,codepage=437,iocharset=utf8,shortname=mixed,errors=remount-ro)
root@shepard:~# 
```

# Works

```
root@shepard:~# mkdir /boot/test1
root@shepard:~# mkdir /boot/TEST1 # This should fail
mkdir: cannot create directory ‘/boot/TEST1’: File exists
root@shepard:~# touch /boot/test1/foo
root@shepard:~# touch /boot/TEST1/bar
root@shepard:~# 
```

# Doesn't work

```
root@shepard:~# mkdir /boot/TEST2
root@shepard:~# mkdir /boot/test2 # This should fail
mkdir: cannot create directory ‘/boot/test2’: File exists
root@shepard:~# touch /boot/TEST2/foo
root@shepard:~# touch /boot/test2/bar
touch: cannot touch '/boot/test2/bar': No such file or directory
root@shepard:~# 
```

**Comment by poettering** [comment]
that sounds very broken and like a kernel bug really...

**Comment by AlexanderKurtz** [comment]
@AlexanderKurtz wrote:

> ```
> root@shepard:~# mkdir /boot/TEST2
> root@shepard:~# mkdir /boot/test2 # This should fail
> mkdir: cannot create directory ‘/boot/test2’: File exists
> root@shepard:~# touch /boot/TEST2/foo
> root@shepard:~# touch /boot/test2/bar
> touch: cannot touch '/boot/test2/bar': No such file or directory
> root@shepard:~# 
> ```

@poettering wrote:

> that sounds very broken and like a kernel bug really...

It's certainly confusing, but I don't think it's a bug. `mount(8)` says:

> ```
>   shortname=mode
>   [...]
>          mixed  Display the short name as is; store a long name when  the
>                 short  name  is  not  all  upper  case.  This mode is the
>                 default since Linux 2.6.32.
> ```

So with `shortname == mixed` there is no long name entry for `TEST2`. And I **guess** that the kernel interpretes `test2` as being a long name (because it would also create it as such), so it looks for a file with the long name `test2`, which simply doesn't exist.

**Comment by evverx** [comment]
> It's certainly confusing, but I don't think it's a bug

Maybe. But

``` sh
# cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux stretch/sid"
NAME="Debian GNU/Linux"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

# uname -r
4.6.0-1-amd64

# sudo stap -v /usr/share/doc/systemtap-doc/examples/general/para-callgraph-verbose.stp '{module("fat"),module("vfat")}.function("*")' -c 'touch /tmp/mount/test2/bar'
Pass 1: parsed user script and 113 library scripts using 102884virt/38008res/6232shr/31988data kb, in 90usr/10sys/101real ms.
Pass 2: analyzed script: 284 probes, 803 functions, 4 embeds, 2 globals using 125868virt/62260res/7516shr/54972data kb, in 610usr/60sys/678real ms.
Pass 3: using cached /root/.systemtap/cache/8b/stap_8b8b8ab6e3da965d1da11b21ed01536a_1287589.c
Pass 4: using cached /root/.systemtap/cache/8b/stap_8b8b8ab6e3da965d1da11b21ed01536a_1287589.ko
Pass 5: starting run.
touch: cannot touch '/tmp/mount/test2/bar': No such file or directory
WARNING: Child process exited with status 1
     0 touch(13241):->vfat_hashi dentry={.d_flags=2097159, .d_seq={.sequence=2}, .d_hash={.next=0x0, .pprev=0x0}, .d_parent=0xffff88001c561c00, .d_name={<union>={<class>={.hash=0, .len=1}, .hash_len=4294967296}, .name="/"}, .d_inode=0xffff88001c57e608, .d_iname="/", .d_lockref={<union>={.lock_count=21474836480, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=5}}}, .d_op=0xffffffffc02e3200, .d_sb=0xffff88001bab1800, .d_time=7310590614138020449, .d_fsdata=0x0, .d_lru={.next=0xffff88001c561c80, .prev=0xffff88001c561
    10 touch(13241):<-vfat_hashi return=0x0
     0 touch(13241):->vfat_cmpi parent={.d_flags=2097159, .d_seq={.sequence=2}, .d_hash={.next=0x0, .pprev=0x0}, .d_parent=0xffff88001c561c00, .d_name={<union>={<class>={.hash=0, .len=1}, .hash_len=4294967296}, .name="/"}, .d_inode=0xffff88001c57e608, .d_iname="/", .d_lockref={<union>={.lock_count=21474836480, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=5}}}, .d_op=0xffffffffc02e3200, .d_sb=0xffff88001bab1800, .d_time=7310590614138020449, .d_fsdata=0x0, .d_lru={.next=0xffff88001c561c80, .prev=0xffff88001c561
     5 touch(13241):<-vfat_cmpi return=0x0
     0 touch(13241):->vfat_revalidate_ci dentry={.d_flags=524487, .d_seq={.sequence=0}, .d_hash={.next=0x0, .pprev=0xffffc900000681f8}, .d_parent=0xffff88001c561c00, .d_name={<union>={<class>={.hash=326478328, .len=5}, .hash_len=21801314808}, .name="test2"}, .d_inode=0x0, .d_iname="test2", .d_lockref={<union>={.lock_count=0, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=0}}}, .d_op=0xffffffffc02e3200, .d_sb=0xffff88001bab1800, .d_time=3, .d_fsdata=0x0, .d_lru={.next=0xffff88001cc2ec08, .prev=0xffff88001cc2ec08}, .d_chi
     5 touch(13241):<-vfat_revalidate_ci return=0xfffffffffffffff6
     0 touch(13241):->vfat_revalidate_ci dentry={.d_flags=524487, .d_seq={.sequence=0}, .d_hash={.next=0x0, .pprev=0xffffc900000681f8}, .d_parent=0xffff88001c561c00, .d_name={<union>={<class>={.hash=326478328, .len=5}, .hash_len=21801314808}, .name="test2"}, .d_inode=0x0, .d_iname="test2", .d_lockref={<union>={.lock_count=4294967296, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=1}}}, .d_op=0xffffffffc02e3200, .d_sb=0xffff88001bab1800, .d_time=3, .d_fsdata=0x0, .d_lru={.next=0xffff88001cc2ec08, .prev=0xffff88001cc2ec08
    14 touch(13241): ->vfat_revalidate_shortname dentry={.d_flags=524487, .d_seq={.sequence=0}, .d_hash={.next=0x0, .pprev=0xffffc900000681f8}, .d_parent=0xffff88001c561c00, .d_name={<union>={<class>={.hash=326478328, .len=5}, .hash_len=21801314808}, .name="test2"}, .d_inode=0x0, .d_iname="test2", .d_lockref={<union>={.lock_count=4294967296, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=1}}}, .d_op=0xffffffffc02e3200, .d_sb=0xffff88001bab1800, .d_time=3, .d_fsdata=0x0, .d_lru={.next=0xffff88001cc2ec08, .prev=0xffff88001cc2ec08
    18 touch(13241): <-vfat_revalidate_shortname return=0x1
    20 touch(13241):<-vfat_revalidate_ci return=0x1
     0 touch(13241):->vfat_hashi dentry={.d_flags=2097159, .d_seq={.sequence=2}, .d_hash={.next=0x0, .pprev=0x0}, .d_parent=0xffff88001c561c00, .d_name={<union>={<class>={.hash=0, .len=1}, .hash_len=4294967296}, .name="/"}, .d_inode=0xffff88001c57e608, .d_iname="/", .d_lockref={<union>={.lock_count=21474836480, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=5}}}, .d_op=0xffffffffc02e3200, .d_sb=0xffff88001bab1800, .d_time=7310590614138020449, .d_fsdata=0x0, .d_lru={.next=0xffff88001c561c80, .prev=0xffff88001c561
     4 touch(13241):<-vfat_hashi return=0x0
     0 touch(13241):->vfat_cmpi parent={.d_flags=2097159, .d_seq={.sequence=2}, .d_hash={.next=0x0, .pprev=0x0}, .d_parent=0xffff88001c561c00, .d_name={<union>={<class>={.hash=0, .len=1}, .hash_len=4294967296}, .name="/"}, .d_inode=0xffff88001c57e608, .d_iname="/", .d_lockref={<union>={.lock_count=21474836480, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=5}}}, .d_op=0xffffffffc02e3200, .d_sb=0xffff88001bab1800, .d_time=7310590614138020449, .d_fsdata=0x0, .d_lru={.next=0xffff88001c561c80, .prev=0xffff88001c561
     4 touch(13241):<-vfat_cmpi return=0x0
     0 touch(13241):->vfat_revalidate_ci dentry={.d_flags=524487, .d_seq={.sequence=0}, .d_hash={.next=0x0, .pprev=0xffffc900000681f8}, .d_parent=0xffff88001c561c00, .d_name={<union>={<class>={.hash=326478328, .len=5}, .hash_len=21801314808}, .name="test2"}, .d_inode=0x0, .d_iname="test2", .d_lockref={<union>={.lock_count=0, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=0}}}, .d_op=0xffffffffc02e3200, .d_sb=0xffff88001bab1800, .d_time=3, .d_fsdata=0x0, .d_lru={.next=0xffff88001cc2ec08, .prev=0xffff88001cc2ec08}, .d_chi
     4 touch(13241):<-vfat_revalidate_ci return=0xfffffffffffffff6
     0 touch(13241):->vfat_revalidate_ci dentry={.d_flags=524487, .d_seq={.sequence=0}, .d_hash={.next=0x0, .pprev=0xffffc900000681f8}, .d_parent=0xffff88001c561c00, .d_name={<union>={<class>={.hash=326478328, .len=5}, .hash_len=21801314808}, .name="test2"}, .d_inode=0x0, .d_iname="test2", .d_lockref={<union>={.lock_count=4294967296, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=1}}}, .d_op=0xffffffffc02e3200, .d_sb=0xffff88001bab1800, .d_time=3, .d_fsdata=0x0, .d_lru={.next=0xffff88001cc2ec08, .prev=0xffff88001cc2ec08
    13 touch(13241): ->vfat_revalidate_shortname dentry={.d_flags=524487, .d_seq={.sequence=0}, .d_hash={.next=0x0, .pprev=0xffffc900000681f8}, .d_parent=0xffff88001c561c00, .d_name={<union>={<class>={.hash=326478328, .len=5}, .hash_len=21801314808}, .name="test2"}, .d_inode=0x0, .d_iname="test2", .d_lockref={<union>={.lock_count=4294967296, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=1}}}, .d_op=0xffffffffc02e3200, .d_sb=0xffff88001bab1800, .d_time=3, .d_fsdata=0x0, .d_lru={.next=0xffff88001cc2ec08, .prev=0xffff88001cc2ec08
    17 touch(13241): <-vfat_revalidate_shortname return=0x1
    19 touch(13241):<-vfat_revalidate_ci return=0x1
WARNING: /usr/bin/staprun exited with status: 1
Pass 5: run completed in 0usr/720sys/1119real ms.
Pass 5: run failed.  [man error::pass5]
```

``` sh
$ cat /etc/os-release
NAME=Fedora
VERSION="23 (Cloud Edition)"
ID=fedora
VERSION_ID=23
PRETTY_NAME="Fedora 23 (Cloud Edition)"
ANSI_COLOR="0;34"
CPE_NAME="cpe:/o:fedoraproject:fedora:23"
HOME_URL="https://fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=23
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=23
PRIVACY_POLICY_URL=https://fedoraproject.org/wiki/Legal:PrivacyPolicy
VARIANT="Cloud Edition"
VARIANT_ID=cloud

# uname -r
4.5.7-202.fc23.x86_64

# sudo stap -v /usr/share/doc/systemtap-client/examples/general/para-callgraph-verbose.stp 'module("vfat").function("*")' -c 'touch /tmp/mount/test2/bar'
Pass 1: parsed user script and 230 library scripts using 697968virt/606784res/7380shr/599612data kb, in 2240usr/260sys/2490real ms.
Pass 2: analyzed script: 40 probes, 232 functions, 4 embeds, 2 globals using 737232virt/647552res/8848shr/638876data kb, in 360usr/170sys/540real ms.
Pass 3: using cached /root/.systemtap/cache/eb/stap_eb87938f261ceb9a47f3cbc86f0d66d1_326263.c
Pass 4: using cached /root/.systemtap/cache/eb/stap_eb87938f261ceb9a47f3cbc86f0d66d1_326263.ko
Pass 5: starting run.
     0 touch(5723):->vfat_hashi dentry={.d_flags=2097159, .d_seq={.sequence=2}, .d_hash={.next=0x0, .pprev=0x0}, .d_parent=0xffff8800170520c0, .d_name={<union>={<class>={.hash=0, .len=1}, .hash_len=4294967296}, .name="/"}, .d_inode=0xffff88001ba1d3c8, .d_iname="/", .d_lockref={<union>={.lock_count=17179869184, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=4}}}, .d_op=0xffffffffa00d5200, .d_sb=0xffff88007b273800, .d_time=0, .d_fsdata=0x0, .d_lru={.next=0xffff880017052140, .prev=0xffff880017052140}, .d_child={.n
     7 touch(5723):<-vfat_hashi return=0x0
     0 touch(5723):->vfat_cmpi parent={.d_flags=2097159, .d_seq={.sequence=2}, .d_hash={.next=0x0, .pprev=0x0}, .d_parent=0xffff8800170520c0, .d_name={<union>={<class>={.hash=0, .len=1}, .hash_len=4294967296}, .name="/"}, .d_inode=0xffff88001ba1d3c8, .d_iname="/", .d_lockref={<union>={.lock_count=17179869184, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=4}}}, .d_op=0xffffffffa00d5200, .d_sb=0xffff88007b273800, .d_time=0, .d_fsdata=0x0, .d_lru={.next=0xffff880017052140, .prev=0xffff880017052140}, .d_child={.n
     3 touch(5723):<-vfat_cmpi return=0x0
     0 touch(5723):->vfat_revalidate_ci dentry={.d_flags=2621639, .d_seq={.sequence=4}, .d_hash={.next=0xffff880051df69c8, .pprev=0xffffc900001d87d0}, .d_parent=0xffff8800170520c0, .d_name={<union>={<class>={.hash=326478328, .len=5}, .hash_len=21801314808}, .name="TEST2"}, .d_inode=0xffff88001ba1d688, .d_iname="TEST2", .d_lockref={<union>={.lock_count=8589934592, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=2}}}, .d_op=0xffffffffa00d5200, .d_sb=0xffff88007b273800, .d_time=2, .d_fsdata=0x0, .d_lru={.next=0xffff8800171
     3 touch(5723):<-vfat_revalidate_ci return=0xfffffffffffffff6
     0 touch(5723):->vfat_revalidate_ci dentry={.d_flags=2621639, .d_seq={.sequence=4}, .d_hash={.next=0xffff880051df69c8, .pprev=0xffffc900001d87d0}, .d_parent=0xffff8800170520c0, .d_name={<union>={<class>={.hash=326478328, .len=5}, .hash_len=21801314808}, .name="TEST2"}, .d_inode=0xffff88001ba1d688, .d_iname="TEST2", .d_lockref={<union>={.lock_count=12884901888, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=3}}}, .d_op=0xffffffffa00d5200, .d_sb=0xffff88007b273800, .d_time=2, .d_fsdata=0x0, .d_lru={.next=0xffff880017
     3 touch(5723):<-vfat_revalidate_ci return=0x1
     0 touch(5723):->vfat_hashi dentry={.d_flags=2621639, .d_seq={.sequence=4}, .d_hash={.next=0xffff880051df69c8, .pprev=0xffffc900001d87d0}, .d_parent=0xffff8800170520c0, .d_name={<union>={<class>={.hash=326478328, .len=5}, .hash_len=21801314808}, .name="TEST2"}, .d_inode=0xffff88001ba1d688, .d_iname="TEST2", .d_lockref={<union>={.lock_count=12884901888, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=3}}}, .d_op=0xffffffffa00d5200, .d_sb=0xffff88007b273800, .d_time=2, .d_fsdata=0x0, .d_lru={.next=0xffff880017
     2 touch(5723):<-vfat_hashi return=0x0
     0 touch(5723):->vfat_cmpi parent={.d_flags=2621639, .d_seq={.sequence=4}, .d_hash={.next=0xffff880051df69c8, .pprev=0xffffc900001d87d0}, .d_parent=0xffff8800170520c0, .d_name={<union>={<class>={.hash=326478328, .len=5}, .hash_len=21801314808}, .name="TEST2"}, .d_inode=0xffff88001ba1d688, .d_iname="TEST2", .d_lockref={<union>={.lock_count=12884901888, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=3}}}, .d_op=0xffffffffa00d5200, .d_sb=0xffff88007b273800, .d_time=2, .d_fsdata=0x0, .d_lru={.next=0xffff880017
     2 touch(5723):<-vfat_cmpi return=0x0
     0 touch(5723):->vfat_revalidate_ci dentry={.d_flags=524487, .d_seq={.sequence=6}, .d_hash={.next=0x0, .pprev=0xffffc9000010d080}, .d_parent=0xffff88005fbb2240, .d_name={<union>={<class>={.hash=2303653, .len=3}, .hash_len=12887205541}, .name="bar"}, .d_inode=0x0, .d_iname="bar", .d_lockref={<union>={.lock_count=4294967296, <class>={.lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, .count=1}}}, .d_op=0xffffffffa00d5200, .d_sb=0xffff88007b273800, .d_time=8, .d_fsdata=0x0, .d_lru={.next=0xffff8800363f9c08, .prev=0xffff88001712f8c0}, .d_
     2 touch(5723):<-vfat_revalidate_ci return=0x0
     0 touch(5723):->vfat_lookup dir={.i_mode=16877, .i_opflags=3, .i_uid={.val=0}, .i_gid={.val=0}, .i_flags=0, .i_acl=0xffffffffffffffff, .i_default_acl=0xffffffffffffffff, .i_op=0xffffffffa00d5080, .i_sb=0xffff88007b273800, .i_mapping=0xffff88001ba1d7f0, .i_security=0xffff88007a4fe780, .i_ino=12, <union>={.i_nlink=2, .__i_nlink=2}, .i_rdev=0, .i_size=2048, .i_atime={.tv_sec=1468948550, .tv_nsec=0}, .i_mtime={.tv_sec=1468948550, .tv_nsec=0}, .i_ctime={.tv_sec=1468947746, .tv_nsec=0}, .i_lock={<union>={.rlock={.raw_lock={.val={.counter=0
    11 touch(5723): ->vfat_find dir={.i_mode=16877, .i_opflags=3, .i_uid={.val=0}, .i_gid={.val=0}, .i_flags=0, .i_acl=0xffffffffffffffff, .i_default_acl=0xffffffffffffffff, .i_op=0xffffffffa00d5080, .i_sb=0xffff88007b273800, .i_mapping=0xffff88001ba1d7f0, .i_security=0xffff88007a4fe780, .i_ino=12, <union>={.i_nlink=2, .__i_nlink=2}, .i_rdev=0, .i_size=2048, .i_atime={.tv_sec=1468948550, .tv_nsec=0}, .i_mtime={.tv_sec=1468948550, .tv_nsec=0}, .i_ctime={.tv_sec=1468947746, .tv_nsec=0}, .i_lock={<union>={.rlock={.raw_lock={.val={.counter=0
    21 touch(5723): <-vfat_find return=0xfffffffffffffffe
    23 touch(5723):<-vfat_lookup return=0x0
     0 touch(5723):->vfat_create dir={.i_mode=16877, .i_opflags=3, .i_uid={.val=0}, .i_gid={.val=0}, .i_flags=0, .i_acl=0xffffffffffffffff, .i_default_acl=0xffffffffffffffff, .i_op=0xffffffffa00d5080, .i_sb=0xffff88007b273800, .i_mapping=0xffff88001ba1d7f0, .i_security=0xffff88007a4fe780, .i_ino=12, <union>={.i_nlink=2, .__i_nlink=2}, .i_rdev=0, .i_size=2048, .i_atime={.tv_sec=1468948550, .tv_nsec=0}, .i_mtime={.tv_sec=1468948550, .tv_nsec=0}, .i_ctime={.tv_sec=1468947746, .tv_nsec=0}, .i_lock={<union>={.rlock={.raw_lock={.val={.counter=0
    14 touch(5723): ->vfat_add_entry dir={.i_mode=16877, .i_opflags=3, .i_uid={.val=0}, .i_gid={.val=0}, .i_flags=0, .i_acl=0xffffffffffffffff, .i_default_acl=0xffffffffffffffff, .i_op=0xffffffffa00d5080, .i_sb=0xffff88007b273800, .i_mapping=0xffff88001ba1d7f0, .i_security=0xffff88007a4fe780, .i_ino=12, <union>={.i_nlink=2, .__i_nlink=2}, .i_rdev=0, .i_size=2048, .i_atime={.tv_sec=1468948550, .tv_nsec=0}, .i_mtime={.tv_sec=1468948550, .tv_nsec=0}, .i_ctime={.tv_sec=1468947746, .tv_nsec=0}, .i_lock={<union>={.rlock={.raw_lock={.val={.counter=0
    25 touch(5723):  ->vfat_find_form dir={.i_mode=16877, .i_opflags=3, .i_uid={.val=0}, .i_gid={.val=0}, .i_flags=0, .i_acl=0xffffffffffffffff, .i_default_acl=0xffffffffffffffff, .i_op=0xffffffffa00d5080, .i_sb=0xffff88007b273800, .i_mapping=0xffff88001ba1d7f0, .i_security=0xffff88007a4fe780, .i_ino=12, <union>={.i_nlink=2, .__i_nlink=2}, .i_rdev=0, .i_size=2048, .i_atime={.tv_sec=1468948550, .tv_nsec=0}, .i_mtime={.tv_sec=1468948550, .tv_nsec=0}, .i_ctime={.tv_sec=1468947746, .tv_nsec=0}, .i_lock={<union>={.rlock={.raw_lock={.val={.counter=0
    29 touch(5723):  <-vfat_find_form return=0xfffffffffffffffe
    32 touch(5723): <-vfat_add_entry return=0x0
    37 touch(5723):<-vfat_create return=0x0
Pass 5: run completed in 0usr/210sys/553real ms.
```

**Comment by AlexanderKurtz** [comment]
http://pkgs.fedoraproject.org/cgit/rpms/kernel.git/tree/config-generic#n4612

vs.

https://anonscm.debian.org/cgit/kernel/linux.git/tree/debian/config/config#n5109

# Verification

```
root@shepard:~# umount /boot/
root@shepard:~# mount /dev/sda2 /boot
root@shepard:~# dmesg | tail -1
[34273.481849] FAT-fs (sda2): utf8 is not a recommended IO charset for FAT filesystems, filesystem will be case sensitive!
root@shepard:~# touch /boot/TEST
root@shepard:~# file /boot/Test
/boot/Test: cannot open `/boot/Test' (No such file or directory)
root@shepard:~# 
root@shepard:~# umount /boot/
root@shepard:~# mount -o iocharset=ascii /dev/sda2 /boot
root@shepard:~# dmesg | tail -1
[34273.481849] FAT-fs (sda2): utf8 is not a recommended IO charset for FAT filesystems, filesystem will be case sensitive!
root@shepard:~# # This is the warning from the first mount, compare the time stamps
root@shepard:~# touch /boot/TEST
root@shepard:~# file /boot/Test
/boot/Test: empty
root@shepard:~# 
```

**Comment by AlexanderKurtz** [comment]
So, to summarize:

With `iocharset == ascii` (the Fedora default), a file created as `FooBarBaz` can also be accessed as `foobarbaz` (and vice versa), while with `iorcharset == utf8` (the Debian default, which causes the FS to become case-sensitive) they are two different files.

So far, so easy. However:

With `iocharset == ascii`, a file created as `FooBar` can also be accessed as `foobar`, while with `iorcharset == utf8`, they are two different files **of which only one may exist**, because both will try to use the same short name `FOOBAR`.

The difference between the the cases? The freaking length of the file names...

**Comment by AlexanderKurtz** [comment]
Corollary: A file created as `FooBar` (or any other case variant!) can always be accessed using its 8.3 short name `FOOBAR`, regardless of the values of the `iocharset=` and `shortname=` mount options.

This suggests a possible solution: Since all the file names defined by the spec [0] are either valid 8.3 short names, or (in the case of `BOOTx64.EFI`) have a 8.3 short name without the infamous number suffix (`FOOBAR~1.TXT`), simply always using the 8.3 short name should™work.

**TL;DR:** Always use upper case only, for all EFI files/directories.

[0]  [UEFI Specification](http://www.uefi.org/specifications), version 2.6, section 3.5.1.1

**Comment by AlexanderKurtz** [comment]
Fix waiting in #3770.

**Comment by evverx** [comment]
> CONFIG_FAT_DEFAULT_IOCHARSET="ascii" vs CONFIG_FAT_DEFAULT_IOCHARSET="utf8"

Oh, right. Thanks!
So, the reproducer:

``` sh
# 4.5.7-202.fc23.x86_64

dd if=/dev/zero bs=1M count=1 of=/var/tmp/img
mkfs.vfat /var/tmp/img
mkdir /tmp/mount

mount -o iocharset=utf8 /var/tmp/img /tmp/mount/

touch /tmp/mount/HEY # ok
file /tmp/mount/HEY # /tmp/mount/HEY: empty

touch /tmp/mount/hey # touch: cannot touch ‘/tmp/mount/hey’: File exists
file /tmp/mount/hey # /tmp/mount/hey: cannot open `/tmp/mount/hey' (No such file or directory)
```

> while with iorcharset == utf8, they are two different files of which only one may exist, because both will try to use the same short name FOOBAR.

``` sh
dd if=/dev/zero bs=1M count=1 of=/var/tmp/img
mkfs.vfat /var/tmp/img
mkdir /tmp/mount

mount -o iocharset=utf8 /var/tmp/img /tmp/mount/

touch /tmp/mount/hey
file /tmp/mount/hey # /tmp/mount/hey: empty

touch /tmp/mount/HEY
file /tmp/mount/HEY # /tmp/mount/HEY: empty
```

Am I missing something?

**Comment by evverx** [comment]
BTW:
This is a known issue: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=745280#37
Was fixed by https://bugs.debian.org/cgi-bin/bugreport.cgi?att=1;bug=745280;filename=0001-setup-Normalize-the-boot-.efi-to-the-case-used-by-th.patch;msg=47
